### ISSUE TITLE
[Metal] partial writes to a torch/numpy array zero the untouched elements

### ISSUE THREAD

**Issue by tlequy-qc** [Genesis-Embodied-AI/quadrants#841]
**Describe the bug**

On the Metal backend, passing a `torch.Tensor` (or `numpy.ndarray`) to a kernel that
**writes the array without ever reading it** discards the array's prior contents:
every element the kernel does not write comes back as `0` instead of keeping its
previous value.

A `qd.ndarray` is unaffected. CPU is unaffected. Nothing warns at compile time.

**To Reproduce**

```py
import quadrants as qd
import torch

qd.init(arch=qd.metal)

@qd.kernel
def write_even(out: qd.types.NDArray[qd.i32, 1]) -> None:
    for pid in range(out.shape[0]):
        if pid % 2 == 0:
            out[pid] = 42  # only even lanes write; odd lanes must keep their prior value

qd_out = qd.ndarray(qd.i32, (8,))
qd_out.fill(7)
torch_out = torch.full((8,), 7, dtype=torch.int32)

write_even(torch_out)
write_even(qd_out)

print(torch_out)                # want [42,7,42,7,...]
print(qd_out.to_torch().cpu())  # want [42,7,42,7,...]
```

**Log/Screenshots**

```
[Quadrants] version 1.2.0, llvm 22.1.0, commit 3d9af718, osx, python 3.12.13
[Quadrants] Starting on arch=metal
tensor([42,  0, 42,  0, 42,  0, 42,  0], dtype=torch.int32)   # torch: prior 7s wiped
tensor([42,  7, 42,  7, 42,  7, 42,  7], dtype=torch.int32)   # qd.ndarray: correct
```

---

**Additional Comments (Opus 5 under human supervision)**

The trigger is not the conditional itself but the resulting access pattern —
"stored to, never loaded from" — so it also fires on an early `continue` or an
`ndrange` that does not cover the whole array.

Scope:

| kernel body, array pre-filled with `7` | `arch=metal` | `arch=cpu` |
| --- | --- | --- |
| `if pid % 2 == 0: out[pid] = 42` | `[42,0,42,0,…]` ✗ | `[42,7,42,7,…]` ✓ |
| `out[pid] = out[pid] + 1` (real load) | `[8,8,…]` ✓ | ✓ |
| `sink[0] += out[pid]` (array only read) | untouched ✓ | ✓ |
| same kernel, `qd.ndarray` instead | ✓ | ✓ |

`numpy.ndarray` behaves exactly like `torch.Tensor`. The staging buffer is freshly
zeroed per launch rather than pooled, so this is data loss, not a leak of another
array's contents (checked by priming with a same-size array first).

The gating logic below lives in `runtime/gfx/`, so **Vulkan is expected to be affected
too**; the LLVM CPU/CUDA runtimes do not stage external arrays and so are not. I only
had Metal and CPU available to test.

**Root cause:**

Three sites, all at commit `3d9af718`:

1. `transforms/detect_read_only.cpp:42-81` — `ExternalPtrAccessVisitor` builds a
   per-argument access mask from the **optimized** IR: `GlobalLoadStmt` → `READ`,
   `GlobalStoreStmt` → `WRITE`, `AtomicOpStmt` → `READ|WRITE`
   (`ExternalPtrAccess{NONE=0, READ=1, WRITE=2}`, `ir/transforms.h:123`).
2. `runtime/gfx/runtime.cpp:511` — the per-launch staging buffer is allocated with
   `host_write = access & READ`.
3. `runtime/gfx/runtime.cpp:78` — the host→device blit is gated on `access & READ`,
   while the device→host readback at `runtime/gfx/runtime.cpp:180` is gated on
   `access & WRITE`.

A write-only argument gets mask `WRITE`, so step 3 skips the upload: the kernel runs
against a freshly allocated (zero) device buffer, and the readback blits that whole
buffer back over the user's array. The elision assumes *no reads ⇒ the kernel
overwrites every element*, which any partial write breaks.

### Exact IR

**Failing kernel** — `if pid % 2 == 0: out[pid] = 42`

```
kernel {
  $0 = offloaded range_for(0, $5) grid_dim=0 block_dim=128
  body {
    <i32> $1 = const 42
    <i32> $2 = const 0
    <i32> $3 = const 1
    <i32> $4 = const 2
    <i32> $5 = external_tensor_shape_along_axis 0, arg_id [0]
    <i32> $6 = loop $0 index 0
    <i32> $7 = div $6 $4
    <u1> $8 = cmp_lt $6 $2
    <u1> $9 = const false
    <i32> $10 = bit_shl $7 $3
    <u1> $11 = cmp_ne $8 $9
    <u1> $12 = cmp_ne $6 $2
    <u1> $13 = cmp_ne $10 $6
    <u1> $14 = logical_and $11 $12
    <u1> $15 = logical_and $14 $13
    <i32> $16 = cast_value<i32> $15
    <i32> $17 = sub $7 $16
    <i32> $18 = bit_shl $17 $3
    <i32> $19 = sub $6 $18          # $19 = pid % 2, via floor-div correction
    <u1> $20 = cmp_eq $19 $2        # $20 = (pid % 2 == 0)
    $21 : if $20 {
      <*struct[none]{0(shape, at 0B): struct[none]{0(dim_0, at 0B): i32}, 1(data_ptr, at 0B): *i32}> $22 = argaddr[0]
      <*i32> $23 = external_ptr $22, [$6] layout=AOS is_grad=false
      $24 : global store [$23 <- $1]     # <-- the ONLY access to arg 0: a store
    }
  }
}
```

`external_ptr $22` is reached by exactly one statement, `global store`, and by no
`global load`. So `arr_access[[0]] == WRITE`, the `access & READ` test at
`runtime/gfx/runtime.cpp:78` is false, the upload is skipped — and the `else` lanes
read back whatever the fresh allocation held, i.e. `0`.

Note that `external_ptr` sits **inside** the `if` here: with nothing to hoist it for,
the pointer materializes only on the storing path.

**Working kernel** — same thing plus an explicit read-modify-write on the other branch

```py
if pid % 2 == 0:
    out[pid] = 42
else:
    out[pid] = out[pid]
```

```
kernel {
  $0 = offloaded range_for(0, $5) grid_dim=0 block_dim=128
  body {
    ... identical $1-$20 ...
    <*struct[none]{0(shape, at 0B): struct[none]{0(dim_0, at 0B): i32}, 1(data_ptr, at 0B): *i32}> $21 = argaddr[0]
    <*i32> $22 = external_ptr $21, [$6] layout=AOS is_grad=false   # hoisted out of the if
    $23 : if $20 {
      $24 : global store [$22 <- $1]
    } else {
      <i32> $25 = global load $22        # <-- this load is the only reason it works
      $26 : global store [$22 <- $25]
    }
  }
}
```

The surviving `global load` sets the `READ` bit, the upload happens, and the result is
the expected `[42,7,42,7,…]`.

**A trap worth knowing:** the load must survive optimization. Writing
`out[pid] = out[pid] * 0 + 42` looks like a read-modify-write but the `* 0` is
constant-folded, the load dies, the mask drops back to `WRITE`, and the corruption
returns. That makes any `x[i] = x[i]` workaround silently dependent on no future pass
eliminating self-stores.

### Suggested fix

The identical hazard was already found and fixed for the **gradient** slot a few lines
away. `runtime/gfx/runtime.cpp:100-105` says:

> WRITE covers plain non-atomic partial stores (`x.grad[i] = val` on a torch/numpy
> tensor) - without the mirror, the device buffer would retain allocator garbage at
> indices the kernel did not touch and the symmetric d2h readback below would silently
> overwrite the user's host `.grad` with that garbage.

That path therefore gates on `kGradReadWrite` (`READ|WRITE`,
`runtime/gfx/runtime.cpp:110-113`). The primary data path never got the same
treatment. Restoring the symmetry is two lines:

```cpp
// runtime/gfx/runtime.cpp:78
if (access & (uint32_t(irpass::ExternalPtrAccess::READ) |
              uint32_t(irpass::ExternalPtrAccess::WRITE))) {

// runtime/gfx/runtime.cpp:511
bool host_write = access & (uint32_t(irpass::ExternalPtrAccess::READ) |
                           uint32_t(irpass::ExternalPtrAccess::WRITE));
```

Line 511 is not optional: without it the `device_->map()` in the blit fails with the
runtime's own *"cannot map memory … ensure your memory is allocated with
`host_read=true` or `host_write=true`"*. `host_read` can stay `false` — the readback at
`runtime/gfx/runtime.cpp:221` goes through `readback_data`, which brings its own
staging buffer.

**Cost:** one extra host→device `memcpy` per launch for write-only external arguments.
Measured on an M-series laptop, 16 MiB `i32`, 20 reps, comparing a pure-write kernel
(no blit) against the same kernel with a surviving load (blit): 2.44 → 3.70 ms/launch,
i.e. ≈1.3 ms per 16 MiB (~12.6 GB/s). `qd.ndarray` arguments never enter this path, so
nothing that already avoids external arrays regresses.

If that overhead matters for hot write-only outputs, the follow-up should be an
explicit opt-out (an `assume_full_write`-style argument annotation) rather than a
static "does this store cover every index" analysis, which is not tractable in general.

Failing that, the compiler should at least warn when it elides the upload for an
argument whose stores are not provably unconditional — the current silence is what
makes this cost real debugging time.
